rohmu/delta/snapshot.py

```diff
@@ -74,38 +75,58 @@ def _list_files(self, basepath: Path):
 
         return sorted(result_files)
 
-    def _list_dirs_and_files(self, basepath: Path):
+    def _list_dirs_and_files(self, basepath: Path) -> Tuple[List[Path], List[Path]]:
         files = self._list_files(basepath)
         dirs = {p.parent for p in files}
         return sorted(dirs), files
 
-    def _add_snapshotfile(self, snapshotfile: SnapshotFile):
+    def _add_snapshotfile(self, snapshotfile: SnapshotFile) -> None:
         old_snapshotfile = self.relative_path_to_snapshotfile.get(snapshotfile.relative_path, None)
         if old_snapshotfile:
             self._remove_snapshotfile(old_snapshotfile)
         self.relative_path_to_snapshotfile[snapshotfile.relative_path] = snapshotfile
         if snapshotfile.hexdigest:
             self.hexdigest_to_snapshotfiles.setdefault(snapshotfile.hexdigest, []).append(snapshotfile)
 
-    def _remove_snapshotfile(self, snapshotfile: SnapshotFile):
+    def _remove_snapshotfile(self, snapshotfile: SnapshotFile) -> None:
         assert self.relative_path_to_snapshotfile[snapshotfile.relative_path] == snapshotfile
         del self.relative_path_to_snapshotfile[snapshotfile.relative_path]
         if snapshotfile.hexdigest:
             self.hexdigest_to_snapshotfiles[snapshotfile.hexdigest].remove(snapshotfile)
 
-    def _snapshotfile_from_path(self, relative_path):
+    def _snapshotfile_from_path(self, relative_path: Path, missing_ok: bool) -> SnapshotFile:
         src_path = self.src / relative_path
         st = src_path.stat()
-        return SnapshotFile(relative_path=relative_path, mtime_ns=st.st_mtime_ns, file_size=st.st_size, stored_file_size=0)
+        return SnapshotFile(
+            relative_path=relative_path,
+            mtime_ns=st.st_mtime_ns,
+            file_size=st.st_size,
+            stored_file_size=0,
+            missing_ok=missing_ok,
+        )
 
-    def _gen_snapshot_hashes(self, relative_paths, reuse_old_snapshotfiles):
+    def _gen_snapshot_hashes(
+        self, relative_paths: Sequence[Path], reuse_old_snapshotfiles: bool, required_paths: Optional[Set[Path]] = None
```

Comment:
nit:
do we need required_paths to be an optional set? Can't we make this just a set, and if it's empty... it's empty? (we can use a frozenset as default) I think we could remove some ifs down there. (if required paths is an empty set, missing_files will always be an empty set, et cetera).

I think this will make the code simpler and we don't need to check/cast at every line.

Reply:
I agree to an extent, I don't think it will make code much simpler, e.g. this check will not go away
```
        if required_paths:
            missing_files = required_paths.difference(set(relative_paths))
```
as I don't want to waste resources on transforming paths to set and calculating the difference between sets if nothing is required.

Reply:
Well, it's a small thing, but what I meant is that `required_paths.difference(set(relative_paths))` in this case will need to be done even if there is no required files, which actually means transforming relative_paths to set for no reason, if the list is big - unnecessary extra memory usage.